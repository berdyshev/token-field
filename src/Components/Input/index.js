import DateTimeInput from './DateTimeInput';
import DurationInput from './DurationInput';
import NumericRangeInput from './NumericRangeInput';
import SelectInput from './SelectInput';
import TextInput from './TextInput';

export {
  DateTimeInput,
  DurationInput,
  NumericRangeInput,
  SelectInput,
  TextInput,
};
