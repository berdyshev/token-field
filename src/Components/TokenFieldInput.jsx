import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import AutosizeInput from 'react-input-autosize';
import Highlighter from 'react-highlight-words';
import { List, ListElement, Scrollbar, Popper } from 'deskpro-components/lib/Components/Common';
import styles from '../styles/style.css';

export default class TokenFieldInput extends React.Component {
  static propTypes = {
    tokenTypes:          PropTypes.arrayOf(PropTypes.object).isRequired,
    tokenKey:            PropTypes.number.isRequired,
    onChange:            PropTypes.func,
    addToken:            PropTypes.func.isRequired,
    selectPreviousToken: PropTypes.func.isRequired,
    selectNextToken:     PropTypes.func.isRequired,
    removeToken:         PropTypes.func.isRequired,
    value:               PropTypes.string.isRequired,
  };

  static defaultProps = {
    onChange() {},
  };

  constructor(props) {
    super(props);
    this.state = {
      value:         props.value,
      tokenKey:      props.tokenKey,
      tokens:        [],
      keyword:       '',
      selectedToken: null,
    };
  }

  focus() {
    this.input.focus();
  }

  selectToken(token) {
    let value = '';
    const { tokenKey } = this.state;
    const match = this.state.value.match(/(.*) [-a-z:]{2,}$/);
    if (match) {
      value = match[1];
    }
    if (value) {
      this.setState({
        value,
        tokenKey,
        tokens: [],
      }, () => {
        this.props.addToken(tokenKey + 1, token.id);
        this.input.blur();
      });
    } else {
      this.props.removeToken(tokenKey);
      this.props.addToken(tokenKey, token.id);
    }
    this.closePopper();
  }

  openPopper = () => {
    this.popperRef.open();
  };

  closePopper = () => {
    this.popperRef.close();
    this.setState({
      tokens: []
    });
  };

  handleBlur = () => {
    if (this.state.value !== '') {
      this.props.onChange(this.state.tokenKey, this.state.value);
      this.closePopper();
    } else {
      this.props.removeToken(this.state.tokenKey);
    }
  };

  handleChange = (event) => {
    let { selectedToken } = this.state;
    const value = event.currentTarget.value;
    const match = value.match(/ ?([-a-z:]{2,})$/);
    let tokens = [];
    let keyword = '';
    if (match) {
      keyword = match[1];
      tokens = this.props.tokenTypes.filter(token =>
        token.id.match(keyword)
      );
    }
    if (tokens.length) {
      if (!selectedToken || tokens.indexOf(selectedToken) === -1) {
        selectedToken = tokens[0];
      }
      this.openPopper();
    } else {
      selectedToken = null;
      this.closePopper();
    }
    this.setState({
      value,
      tokens,
      keyword,
      selectedToken,
    });
  };

  handleKeyDown = (e) => {
    const { tokens } = this.state;
    if (this.state.tokens.length > 0) {
      let index = 0;
      if (['ArrowDown', 'ArrowUp'].indexOf(e.key) !== -1) {
        index = tokens.findIndex(token => token.id === this.state.selectedToken.id);
      }
      switch (e.key) {
        case 'ArrowDown':
          if (index < tokens.length - 1) {
            this.setState({
              selectedToken: tokens[index + 1]
            });
          }
          break;
        case 'ArrowUp':
          if (index > 0) {
            this.setState({
              selectedToken: tokens[index - 1]
            });
          }
          break;
        case 'Tab':
          if (e.shiftKey) {
            this.props.selectPreviousToken();
          } else {
            this.selectToken(this.state.selectedToken);
          }
          break;
        case 'Enter':
          this.selectToken(this.state.selectedToken);
          break;
        case ':': {
          let value = '';
          const match = this.state.value.match(/ ?([-a-z:]{2,})$/);
          if (match) {
            value = match[1];
          }
          const token = tokens.find(t => t.id === value);
          if (token) {
            this.selectToken(token);
          }
          break;
        }
        default:
          // Do nothing
          return true;
      }
      e.preventDefault();
      return false;
    }
    switch (e.key) {
      case 'Backspace':
        if (this.state.tokenKey > 0 && this.state.value === '') {
          this.props.removeToken(this.state.tokenKey - 1, this.state.tokenKey - 1);
        } else {
          return true;
        }
        break;
      case 'Tab':
        if (e.shiftKey) {
          this.props.selectPreviousToken();
        } else {
          this.props.selectNextToken();
        }
        break;
      case 'Enter':
        this.props.selectNextToken();
        break;
      default:
        return true;
    }
    e.preventDefault();
    this.input.blur();
    return true;
  };

  renderTokens() {
    const { keyword } = this.state;
    return (
      <div className="dp-select">
        <div className="dp-select__content">
          <Scrollbar>
            <List className="dp-selectable-list">
              {this.state.tokens.map((token) => {
                const selected = (token === this.state.selectedToken) ? styles.selected : '';
                return (
                  <ListElement
                    key={token.id}
                    onClick={() => this.selectToken(token)}
                    className={classNames(styles['token-suggestion'], selected)}
                  >
                    <Highlighter
                      highlightClassName={styles.highlight}
                      searchWords={[keyword]}
                      textToHighlight={token.id}
                    />:&nbsp;
                    <span className={styles.description}>{token.description}</span>
                  </ListElement>
                );
              }
              )}
            </List>
          </Scrollbar>
        </div>
      </div>
    );
  }

  render() {
    const { value } = this.state;
    return (
      <div
        ref={ref => (this.rootRef = ref)}
        style={{ display: 'inline-block' }}
      >
        <AutosizeInput
          ref={(c) => { this.input = c; }}
          inputClassName={styles['raw-text']}
          value={value}
          style={{ fontSize: 14 }}
          onBlur={this.handleBlur}
          onChange={this.handleChange}
          onKeyDown={this.handleKeyDown}
        />
        <Popper
          ref={ref => (this.popperRef = ref)}
          target={this.rootRef}
          placement="bottom"
          arrow={false}
          onOpen={this.handlePopperOpen}
          onClose={this.handlePopperClose}
        >
          {this.renderTokens()}
        </Popper>
      </div>
    );
  }
}
