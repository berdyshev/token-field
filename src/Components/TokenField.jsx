import React from 'react';
import PropTypes from 'prop-types';
import Icon from 'deskpro-components/lib/Components/Icon';
import * as inputs from './Input';
import TokenFieldInput from './TokenFieldInput';
import styles from '../styles/style.css';

export class TokenField extends React.Component {
  static propTypes = {
    tokenTypes: PropTypes.arrayOf(PropTypes.shape({
      id:     PropTypes.string.isRequired,
      widget: PropTypes.oneOf(
        ['TextInput', 'DateTimeInput', 'DurationInput', 'SelectInput', 'NumericRangeInput']
      ).isRequired,
      props:       PropTypes.object,
      description: PropTypes.string,
    })).isRequired,
    onChange: PropTypes.func,
    value:    PropTypes.array.isRequired,
  };

  static defaultProps = {
    onChange() {},
  };

  constructor(props) {
    super(props);
    this.state = {
      value: props.value
    };

    this.inputs = [];
  }

  componentDidUpdate() {
    if (this.focusInput !== undefined) {
      this.inputs[this.focusInput].focus();
      this.focusInput = undefined;
    }
  }

  getInputField = (key, value) => (
    <TokenFieldInput
      ref={(c) => { this.inputs[key] = c; }}
      value={value}
      key={key}
      tokenKey={key}
      tokenTypes={this.props.tokenTypes}
      addToken={this.addTokenAndFocus}
      onChange={this.handleTokenChange}
      selectPreviousToken={() => this.selectPreviousToken(key)}
      selectNextToken={() => this.selectNextToken(key)}
      removeToken={this.removeToken}
    />
  );

  addInputAndFocus = (key) => {
    const { value } = this.state;
    const inputKey = (!key) ? value.length : key;
    if (this.inputs[0] === null) {
      setTimeout(() => this.addInputAndFocus(key), 10);
      return;
    }
    if (value.length && inputKey === value.length && value[inputKey - 1].type === 'TEXT') {
      this.inputs[inputKey - 1].focus();
    } else {
      value.push({ type: 'TEXT', value: '' });
      this.focusInput = inputKey;
      this.setState({
        value
      });
      this.props.onChange(value);
    }
  };

  addTokenAndFocus = (key, id) => {
    const { value } = this.state;
    value.splice(key, 0, { type: id, value: undefined });
    this.focusInput = key;
    this.setState({
      value
    });
  };

  removeToken = (key, focusKey) => {
    const { value } = this.state;
    value.splice(key, 1);
    if (focusKey !== undefined) {
      this.focusInput = focusKey;
    }
    this.setState({
      value
    });
    this.props.onChange(value);
  };

  selectPreviousToken(key) {
    if (key > 0) {
      this.inputs[key - 1].focus(true);
    }
  }

  selectNextToken(key) {
    if (this.inputs[key + 1]) {
      this.inputs[key + 1].focus();
    } else {
      this.addInputAndFocus();
    }
  }

  handleTokenChange = (key, token) => {
    const { value } = this.state;
    value[key].value = token;
    this.setState({
      value
    });
    this.props.onChange(value);
  };

  renderInputs() {
    const { tokenTypes } = this.props;
    const { value } = this.state;
    let key = 0;
    const elements = [];
    value.forEach((token) => {
      const index = key;
      if (token.type === 'TEXT') {
        elements.push(this.getInputField(index, token.value));
      } else {
        const input = tokenTypes.find(type => type.id === token.type);
        if (input) {
          const Component = inputs[input.widget];
          elements.push(
            <Component
              key={key}
              token={token}
              ref={(c) => { this.inputs[index] = c; }}
              onChange={v => this.handleTokenChange(index, v)}
              removeToken={() => this.removeToken(index)}
              selectPreviousToken={() => this.selectPreviousToken(index)}
              selectNextToken={() => this.selectNextToken(index)}
              {...input.props}
            />
          );
        }
      }
      key += 1;
    });
    return elements;
  }

  render() {
    return (
      <div className={styles['token-field']}>
        <span><Icon name="search" className={styles.search} /></span>
        <div>
          {this.renderInputs()}
        </div>
        <div onClick={() => this.addInputAndFocus()} className={styles['input-box']} />
      </div>
    );
  }
}
