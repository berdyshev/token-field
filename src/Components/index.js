import DateTimeInput from './Input/DateTimeInput';
import DurationInput from './Input/DurationInput';
import NumericRangeInput from './Input/NumericRangeInput';
import SelectInput from './Input/SelectInput';
import TextInput from './Input/TextInput';
import TokenField from './TokenField';

export {
  DateTimeInput,
  DurationInput,
  NumericRangeInput,
  SelectInput,
  TextInput,
  TokenField,
};

export default TokenField;
