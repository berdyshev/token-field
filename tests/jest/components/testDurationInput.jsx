import React from 'react';
import renderer from 'react-test-renderer';
import { DurationInput } from 'Components/index';
import noop from 'deskpro-components/lib/utils/noop';

it('+++capturing Snapshot of DurationInput', () => {
  const token = {
    type:  'user-waiting',
    value: {
      inputType: 'relative',
      time:      {
        minutes: 15,
        hours:   1,
      },
      op: '=',
    }
  };
  const renderedValue = renderer.create(
    <DurationInput
      token={token}
      className="test-class"
      selectPreviousToken={noop}
      selectNextToken={noop}
      removeToken={noop}
    />
  ).toJSON();
  expect(renderedValue).toMatchSnapshot();
});
it('+++capturing Snapshot of DurationInput', () => {
  const token = {
    type:  'user-waiting',
    value: {},
  };
  const renderedValue = renderer.create(
    <DurationInput
      token={token}
      className="test-class"
      selectPreviousToken={noop}
      selectNextToken={noop}
      removeToken={noop}
    />
  ).toJSON();
  expect(renderedValue).toMatchSnapshot();
});
it('+++capturing Snapshot of DurationInput', () => {
  const token = {
    type:  'date',
    value: {
      inputType: 'relative',
      time:      {
        minutes: 60,
      },
      timeEnd: {
        days: 2,
      },
      op: 'range',
    }
  };

  const renderedValue = renderer.create(
    <DurationInput
      token={token}
      className="test-class"
      locale="fr"
      translations={{ to: 'Ã ' }}
      selectPreviousToken={noop}
      selectNextToken={noop}
      removeToken={noop}
    />
  ).toJSON();
  expect(renderedValue).toMatchSnapshot();
});
