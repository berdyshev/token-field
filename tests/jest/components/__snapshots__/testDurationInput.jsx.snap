// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`+++capturing Snapshot of DurationInput 1`] = `
<div
  className="test-class"
>
  <div
    className="dp-code label"
  >
    user-waiting
    :
  </div>
  <span
    className="value"
    onClick={[Function]}
  >
    an hour
  </span>
  <div
    className="remove"
    onClick={[Function]}
  >
    X
  </div>
</div>
`;

exports[`+++capturing Snapshot of DurationInput 2`] = `
<div
  className="test-class"
>
  <div
    className="dp-code label"
  >
    user-waiting
    :
  </div>
  <span
    className="value"
    onClick={[Function]}
  >
    ________
  </span>
  <div
    className="remove"
    onClick={[Function]}
  >
    X
  </div>
</div>
`;

exports[`+++capturing Snapshot of DurationInput 3`] = `
<div
  className="test-class"
>
  <div
    className="dp-code label"
  >
    date
    :
  </div>
  <span
    className="value"
    onClick={[Function]}
  >
    une heure Ã  2 jours
  </span>
  <div
    className="remove"
    onClick={[Function]}
  >
    X
  </div>
</div>
`;
